### PR TITLE
Nirasto azure table writer

### PR DESCRIPTION
**Pull request by nikita-rastogi**
### Required for all PRs:

- [ ] Signed [CLA](https://influxdata.com/community/cla/).
- [ ] Associated README.md updated.
- [ ] Has appropriate unit tests.

### CODE REVIEW

**Comment by nikita-rastogi** [comment]
Sample table can be seen at:
Table name : WADMetricsPT1MP10DV2S20180310
Storage Account: ladextensionrgdiag526.

File: etc/telegraf.conf [hunk 1]
```diff
@@ -82,26 +82,35 @@
...
 ###############################################################################
 
-# Configuration for influxdb server to send metrics to
-[[outputs.influxdb]]
+# Configuration for azureTableStorage output plugin to write metrics to azure table
+[[outputs.azureTableStorage]]
+ deploymentId = "deploymentId"
+ resourceId = "subscriptionId/resourceGroup/VMScaleset"
+ accountName = "ladextensionrgdiag526"
+ accountKey = "42WqyNltbP/S3rxbJizeelr4D35EUTU7en5QKgRotT6iWXZ7xtspB6j0/u5fs4kDaiheiIL8K9et0mdcBzcPig=="
```
Comment (abhijeetgaiha) [inline etc/telegraf.conf]: Please change this to use a SAS URL instead of the acct key.

File: plugins/aggregators/basicstats/basicstats.go [hunk 2]
```diff
@@ -2,26 +2,31 @@ package basicstats
 
 import (
```
Comment (abhijeetgaiha) [inline plugins/aggregators/basicstats/basicstats.go]: We will have to create our own aggregator plugin, given the extensive changes. We cannot change the behavior for other users so drastically. Please add another plugin and copy whatever code is needed from this file and revert this file to the original.

File: plugins/inputs/all/all.go [hunk 3]
```diff
@@ -19,7 +19,7 @@ import (
 	_ "github.com/influxdata/telegraf/plugins/inputs/disque"
 	_ "github.com/influxdata/telegraf/plugins/inputs/dmcache"
 	_ "github.com/influxdata/telegraf/plugins/inputs/dns_query"
-	_ "github.com/influxdata/telegraf/plugins/inputs/docker"
+	//_"github.com/influxdata/telegraf/plugins/inputs/docker"
```
Comment (abhijeetgaiha) [inline plugins/inputs/all/all.go]: Why was commenting this out necessary?

File: plugins/inputs/zipkin/cmd/thrift_serialize/thrift_serialize.go [hunk 4]
```diff
@@ -32,7 +32,7 @@ import (
 	"log"
 
 	"github.com/apache/thrift/lib/go/thrift"
-	"github.com/openzipkin/zipkin-go-opentracing/_thrift/gen-go/zipkincore"
+	"github.com/openzipkin/zipkin-go-opentracing/thrift/gen-go/zipkincore"
```
Comment (abhijeetgaiha) [inline plugins/inputs/zipkin/cmd/thrift_serialize/thrift_serialize.go]: Why is this necessary?

File: plugins/outputs/all/all.go [hunk 5]
```diff
@@ -16,7 +16,8 @@ import (
 	_ "github.com/influxdata/telegraf/plugins/outputs/kafka"
 	_ "github.com/influxdata/telegraf/plugins/outputs/kinesis"
 	_ "github.com/influxdata/telegraf/plugins/outputs/librato"
-	_ "github.com/influxdata/telegraf/plugins/outputs/mqtt"
+	//_ "github.com/influxdata/telegraf/plugins/outputs/mqtt"
+	_ "github.com/influxdata/telegraf/plugins/outputs/azureTableStorage"
```
Comment (abhijeetgaiha) [inline plugins/outputs/all/all.go]: Let's call is Azure. We won't be adding a separate plugin for every sink.

File: plugins/outputs/kinesis/kinesis.go [hunk 6]
```diff
@@ -200,7 +200,7 @@ func (k *KinesisOutput) getPartitionKey(metric telegraf.Metric) string {
 		case "static":
 			return k.Partition.Key
 		case "random":
-			u := uuid.NewV4()
+			u, _ := uuid.NewV4()
```
Comment (abhijeetgaiha) [inline plugins/outputs/kinesis/kinesis.go]: Why are we making changes in other plugins?

**Review by abhijeetgaiha — CHANGES_REQUESTED** [review]
There's some broad changes to be made in this iteration. I'll review the actual plugin code in more detail in the next iteration.

File: etc/telegraf.conf (hunk 1 above)
Comment (nikita-rastogi) [reply etc/telegraf.conf]: done.

File: plugins/inputs/all/all.go (hunk 3 above)
Comment (nikita-rastogi) [reply plugins/inputs/all/all.go]: there were some errors due to some dependency. So i commented it out for now.

File: plugins/inputs/zipkin/cmd/thrift_serialize/thrift_serialize.go (hunk 4 above)
Comment (nikita-rastogi) [reply plugins/inputs/zipkin/cmd/thrift_serialize/thrift_serialize.go]: it required downloading and installing thrift library.the url with "_thrift" did not work , checked on internet and the url with "thrift" worked. hence i had to make this change.

File: plugins/outputs/all/all.go (hunk 5 above)
Comment (nikita-rastogi) [reply plugins/outputs/all/all.go]: will we be writing code for event hub sink in the same output plugin? I think there should be a different plugin for EH sink as the 2 main functions of output plugin interface are different for table and eh.

File: plugins/outputs/kinesis/kinesis.go (hunk 6 above)
Comment (nikita-rastogi) [reply plugins/outputs/kinesis/kinesis.go]: it was giving error that uuid.NewV4() returns 2 values, so i had to make this change for the code to work.

File: plugins/aggregators/basicstats/basicstats.go (hunk 2 above)
Comment (nikita-rastogi) [reply plugins/aggregators/basicstats/basicstats.go]: Agree. Writing a separate aggregator plugin.

Comment (nikita-rastogi) [reply plugins/aggregators/basicstats/basicstats.go]: reverted the changes done in basicstats.go to the commit from where i forked my branch.

Comment (nikita-rastogi) [reply plugins/aggregators/basicstats/basicstats.go]: added a separate file with the required functionalities.

File: plugins/inputs/all/all.go (hunk 3 above)
Comment (abhijeetgaiha) [reply plugins/inputs/all/all.go]: Before this is merged, it needs to be fixed.

File: plugins/outputs/kinesis/kinesis.go (hunk 6 above)
Comment (abhijeetgaiha) [reply plugins/outputs/kinesis/kinesis.go]: Do you have the version of GO that is used by the mainline? Please check.

File: plugins/outputs/all/all.go (hunk 5 above)
Comment (abhijeetgaiha) [reply plugins/outputs/all/all.go]: Let's talk offline.

File: etc/telegraf.conf (hunk 1 above)
Comment (nikita-rastogi) [reply etc/telegraf.conf]: Sorry, i wrongly interpreted the comment. Will update once the review is implemented.

File: plugins/outputs/all/all.go (hunk 5 above)
Comment (nikita-rastogi) [reply plugins/outputs/all/all.go]: Resolved offline.

File: etc/telegraf.conf (hunk 1 above)
Comment (nikita-rastogi) [reply etc/telegraf.conf]: done.

File: plugins/aggregators/metricaggregates/metricaggregates.go [hunk 7]
```diff
@@ -0,0 +1,265 @@
+package metricaggregates
```
Comment (abhijeetgaiha) [inline plugins/aggregators/metricaggregates/metricaggregates.go]: Rename this to azuremetrics

File: etc/telegraf.conf [hunk 8]
```diff
@@ -66,42 +66,52 @@
...
 ###############################################################################
 
-# Configuration for influxdb server to send metrics to
-[[outputs.influxdb]]
+# Configuration for azuretablestorage output plugin to write metrics to azure table
+[[outputs.azuretablestorage]]
+ deployment_id = "deploymentId"
+ resource_id = "subscriptionId/resourceGroup/VMScaleset"
+ account_name = "ladextensionrgdiag526"
+ sas_url = "https://ladextensionrgdiag526.table.core.windows.net"
```
Comment (abhijeetgaiha) [inline etc/telegraf.conf]: The sas_url can simply be computed from the storage acct name, it's not necessary to add it to config.

File: etc/telegraf.conf [hunk 9]
```diff
@@ -66,42 +66,52 @@
...
+ sas_token="sv=2017-07-29&ss=bt&srt=sco&sp=rwdlacu&se=2019-03-20T19:34:18Z&st=2018-03-19T11:34:18Z&spr=https&sig=tw%2BfX8RJw%2FLd7%2Fv5K1w4b2bOJwBAPcqkUsFqBB7LllQ%3D"
```
Comment (abhijeetgaiha) [inline etc/telegraf.conf]: Don't check in the actual SAS token.

File: plugins/aggregators/metricaggregates/metricaggregates.go [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+
+func (m *MetricAggregates) SampleConfig() string {
+	return sampleConfig
+}
+
+func (m *MetricAggregates) Description() string {
+	return "Keep the aggregate metricAggregates of each metric passing through."
+}
+
+func (m *MetricAggregates) Add(in telegraf.Metric) {
```
Comment (abhijeetgaiha) [inline plugins/aggregators/metricaggregates/metricaggregates.go]: Does this have support for multiple aggregation intervals?

File: plugins/outputs/azuretablestorage/azuretablestorage.go [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+
+		//period is the period which decides when to transfer the aggregated metrics.Its in format "60s"
+		periodStr := metrics[i].Tags()[Period]
+		table := azureTableStorage.PeriodVsTableNameVsTableRef[periodStr].TableRef
+
+		//two rows are written for each metric as Azure table has optimized prefix search only and no index.
+		rowKey1 := UTCTicks_DescendingOrderStr + "_" + encodedCounterName
+		entity = table.GetEntityReference(partitionKey, rowKey1)
+		entity.Properties = props
+		entity.Insert(FullMetadata, nil)
```
Comment (abhijeetgaiha) [inline plugins/outputs/azuretablestorage/azuretablestorage.go]: Error checks needed here.

File: plugins/outputs/azuretablestorage/azuretablestorage.go [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+
+		rowKey2 := encodedCounterName + "_" + UTCTicks_DescendingOrderStr
+		entity = table.GetEntityReference(partitionKey, rowKey2)
+		entity.Properties = props
+		entity.Insert(FullMetadata, nil)
```
Comment (abhijeetgaiha) [inline plugins/outputs/azuretablestorage/azuretablestorage.go]: Error checks needed here.

File: plugins/outputs/azuretablestorage/azuretablestorage.go [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+	var entity *storage.Entity
+	var props map[string]interface{}
+	partitionKey := getPrimaryKey(azureTableStorage.ResourceId)
+
+	// iterate over the list of metrics and create a new entity for each metrics and add to the table.
+	for i, _ := range metrics {
+		props = metrics[i].Fields()
+		UTCTicks_DescendingOrderStr, encodedCounterName := getRowKeyComponents(props[TIMESTAMP].(string), props[CounterName].(string))
+		props[DeploymentId] = azureTableStorage.DeploymentId
+		props[Host], _ = os.Hostname()
```
Comment (abhijeetgaiha) [inline plugins/outputs/azuretablestorage/azuretablestorage.go]: If the second return value is an error, that needs to be checked.

File: plugins/outputs/azuretablestorage/azuretablestorage.go [hunk 14]
```diff
@@ -0,0 +1,259 @@
...
+
+		//period is the period which decides when to transfer the aggregated metrics.Its in format "60s"
+		periodStr := metrics[i].Tags()[Period]
+		table := azureTableStorage.PeriodVsTableNameVsTableRef[periodStr].TableRef
+
+		//two rows are written for each metric as Azure table has optimized prefix search only and no index.
+		rowKey1 := UTCTicks_DescendingOrderStr + "_" + encodedCounterName
+		entity = table.GetEntityReference(partitionKey, rowKey1)
```
Comment (abhijeetgaiha) [inline plugins/outputs/azuretablestorage/azuretablestorage.go]: Error check needed here.

File: plugins/outputs/azuretablestorage/azuretablestorage.go [hunk 15]
```diff
@@ -0,0 +1,259 @@
...
+		entity.Properties = props
+		entity.Insert(FullMetadata, nil)
+
+		rowKey2 := encodedCounterName + "_" + UTCTicks_DescendingOrderStr
+		entity = table.GetEntityReference(partitionKey, rowKey2)
```
Comment (abhijeetgaiha) [inline plugins/outputs/azuretablestorage/azuretablestorage.go]: Error check needed here.

File: plugins/outputs/azuretablestorage/stringconstants.go [hunk 16]
```diff
@@ -0,0 +1,23 @@
+package azuretablestorage
+
+//parts of table name
+const WADMetrics = "WADMetrics"
```
Comment (abhijeetgaiha) [inline plugins/outputs/azuretablestorage/stringconstants.go]: All names should use '_', not camel casing. This is Go convention.

File: plugins/outputs/azuretablestorage/azuretablestorage.go [hunk 17]
```diff
@@ -0,0 +1,259 @@
...
+
+//RETURNS: primary key for the azure table.
+func getPrimaryKey(resourceId string) string {
+	return encodeSpecialCharacterToUTF16(resourceId)
+}
+
+//RETURNS: difference of max value that can be held by time and number of 100 ns in current time.
+func getUTCTicks_DescendingOrder(lastSampleTimestamp string) uint64 {
+
+	currentTime, _ := time.Parse(layout, lastSampleTimestamp)
```
Comment (abhijeetgaiha) [inline plugins/outputs/azuretablestorage/azuretablestorage.go]: Error check second value, do not drop.

File: plugins/outputs/azuretablestorage/azuretablestorage.go [hunk 18]
```diff
@@ -0,0 +1,259 @@
...
+		return er
+	}
+	tableClient := client.GetTableService()
+	azureTableStorage.PeriodVsTableNameVsTableRef =
+		getAzurePeriodVsTableNameVsTableRefMap(azureTableStorage, tableClient)
+
+	for _, tableVsTableRef := range azureTableStorage.PeriodVsTableNameVsTableRef {
+		er := tableVsTableRef.TableRef.Create(30, FullMetadata, nil)
+		if er != nil && strings.Contains(er.Error(), "TableAlreadyExists") {
+			fmt.Println("the table ", tableVsTableRef.TableName, " already exists.")
```
Comment (abhijeetgaiha) [inline plugins/outputs/azuretablestorage/azuretablestorage.go]: Log to file, not to console.

File: plugins/outputs/azuretablestorage/azuretablestorage.go [hunk 19]
```diff
@@ -0,0 +1,259 @@
...
+	}
+	return periodVsTableNameVsTableRef
+}
+
+func (azureTableStorage *AzureTableStorage) Connect() error {
+	client, er := storage.NewAccountSASClientFromEndpointToken(azureTableStorage.SasUrl, azureTableStorage.SasToken)
+	if er != nil {
+		return er
+	}
+	tableClient := client.GetTableService()
```
Comment (abhijeetgaiha) [inline plugins/outputs/azuretablestorage/azuretablestorage.go]: Error check needed.

File: plugins/outputs/azuretablestorage/azuretablestorage.go [hunk 20]
```diff
@@ -0,0 +1,259 @@
...
+	// iterate over the list of metrics and create a new entity for each metrics and add to the table.
+	for i, _ := range metrics {
+		props = metrics[i].Fields()
+		UTCTicks_DescendingOrderStr, encodedCounterName := getRowKeyComponents(props[TIMESTAMP].(string), props[CounterName].(string))
+		props[DeploymentId] = azureTableStorage.DeploymentId
+		props[Host], _ = os.Hostname()
+
+		//period is the period which decides when to transfer the aggregated metrics.Its in format "60s"
+		periodStr := metrics[i].Tags()[Period]
+		table := azureTableStorage.PeriodVsTableNameVsTableRef[periodStr].TableRef
```
Comment (abhijeetgaiha) [inline plugins/outputs/azuretablestorage/azuretablestorage.go]: Error check needed here.

**Review by abhijeetgaiha — CHANGES_REQUESTED** [review]

File: plugins/aggregators/metricaggregates/metricaggregates.go (hunk 7 above)
Comment (nikita-rastogi) [reply plugins/aggregators/metricaggregates/metricaggregates.go]: done.

File: etc/telegraf.conf (hunk 8 above)
Comment (nikita-rastogi) [reply etc/telegraf.conf]: done.

File: etc/telegraf.conf (hunk 9 above)
Comment (nikita-rastogi) [reply etc/telegraf.conf]: ok. will remove it in final checkin.

File: etc/telegraf.conf [hunk 21]
```diff
@@ -736,6 +746,21 @@
 #   ## aggregator and will not get sent to the output plugins.
 #   drop_original = false
 
+ [[aggregators.metricaggregates]]
```
Comment (nikita-rastogi) [inline etc/telegraf.conf]: adding config for metricaggregates 2 times creates 2 instances of the plgin with different periods.

File: plugins/aggregators/metricaggregates/metricaggregates.go (hunk 10 above)
Comment (nikita-rastogi) [reply plugins/aggregators/metricaggregates/metricaggregates.go]: yes. if you write the config for a plugin n number of times in the conf file, then n instances of that plugin are created. 
The implementation is as follows: 
1. the code reads the config and gets the list of pluginNames (list of all input plugins is the list of names following [[inputs.*]]). 
2. then for each plugin name in the list, an instance is created.

Comment (nikita-rastogi) [reply plugins/aggregators/metricaggregates/metricaggregates.go]: please have a look at the config where the config for metricAggregates is writen 2 times.

File: plugins/outputs/azuretablestorage/azuretablestorage.go (hunk 11 above)
Comment (nikita-rastogi) [reply plugins/outputs/azuretablestorage/azuretablestorage.go]: GetEntityReference returns an empty Entity object with the values being passed to the method. As the object is created offline, it can never be null. Please add comments if i am overlooking anything.

File: plugins/outputs/azuretablestorage/azuretablestorage.go (hunk 13 above)
Comment (nikita-rastogi) [reply plugins/outputs/azuretablestorage/azuretablestorage.go]: done.

File: plugins/outputs/azuretablestorage/azuretablestorage.go (hunk 19 above)
Comment (nikita-rastogi) [reply plugins/outputs/azuretablestorage/azuretablestorage.go]: the object is created offline, no call to the service is made, hence can never be null.

File: plugins/outputs/azuretablestorage/azuretablestorage.go (hunk 12 above)
Comment (nikita-rastogi) [reply plugins/outputs/azuretablestorage/azuretablestorage.go]: same as above.

File: plugins/outputs/azuretablestorage/azuretablestorage.go (hunk 11 above)
Comment (nikita-rastogi) [reply plugins/outputs/azuretablestorage/azuretablestorage.go]: implementation of the GetEntityReference method is at https://github.com/Azure/azure-sdk-for-go/blob/5bf28521f7a6f72f2dde3e5a5fb0d27d7a8aeee9/storage/entity.go#L63

File: plugins/outputs/azuretablestorage/azuretablestorage.go (hunk 14 above)
Comment (nikita-rastogi) [reply plugins/outputs/azuretablestorage/azuretablestorage.go]: same reply as previous comment.

File: plugins/outputs/azuretablestorage/azuretablestorage.go (hunk 15 above)
Comment (nikita-rastogi) [reply plugins/outputs/azuretablestorage/azuretablestorage.go]: same reply as previous comment.

File: plugins/outputs/azuretablestorage/azuretablestorage.go (hunk 17 above)
Comment (nikita-rastogi) [reply plugins/outputs/azuretablestorage/azuretablestorage.go]: done.

File: plugins/outputs/azuretablestorage/azuretablestorage.go (hunk 20 above)
Comment (nikita-rastogi) [reply plugins/outputs/azuretablestorage/azuretablestorage.go]: As explained above, tableRef can never be nil, as its object is created offline.

File: plugins/outputs/azuretablestorage/stringconstants.go (hunk 16 above)
Comment (nikita-rastogi) [reply plugins/outputs/azuretablestorage/stringconstants.go]: in the telegraf code, for config keys they are using '_' and inside the code camelCase is used.

File: plugins/outputs/azuretablestorage/azuretablestorage.go (hunk 18 above)
Comment (nikita-rastogi) [reply plugins/outputs/azuretablestorage/azuretablestorage.go]: done.

**Review by nikita-rastogi — COMMENTED** [review]
Discussed the PR with Abhijeet offline and concluded that the changes are good to be merged to master.
Tests Done before final merge:
1. the rows are getting written to the table as expected from the plugin.
2. Unit tests are ok.
